### ISSUE TITLE
Interoperability versus incorporating with/ other python packages

### ISSUE THREAD

**Issue by haesleinhuepf** [kevinyamauchi/morphometrics#2]
Hi @kevinyamauchi ,

I'm just listing some functions and widgets (I'm aware of) in a couple of napari-plugins which produce measurements from labels. Depending on the strategy of `morphometrics` we could either rip out the measurement code from there and collect it here, or we could implement some interoperability stuff on the side of these plugins to be interoperable with `morphometrics`:

* [napari-skimage-regionprops `regionprops()`](https://github.com/haesleinhuepf/napari-skimage-regionprops/blob/master/napari_skimage_regionprops/_regionprops.py#L12)
* [napari-clusters-plotter `region_props_with_neighborhood_data()`](https://github.com/BiAPoL/napari-clusters-plotter/blob/main/napari_clusters_plotter/_measure.py#L200)
* [napari-simpleitk-image-processing `label_statistics()`](https://github.com/haesleinhuepf/napari-simpleitk-image-processing/blob/main/src/napari_simpleitk_image_processing/_simpleitk_image_processing.py#L789)
* [napari-cupy-image-processing `measurements()`](https://github.com/haesleinhuepf/napari-cupy-image-processing/blob/main/napari_cupy_image_processing/_cupy_image_processing.py#L225) (optional, because of tricky dependencies and limited functionality)
* [napari-pyclesperanto-assistant `statistics_of_labeled_pixels()`](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_statistics_of_labeled_pixels.py#L17) (optional, because of tricky dependencies and limited functionality)
* [PartSeg `NapariMeasurementWidget`](https://github.com/4DNucleome/PartSeg/blob/develop/package/PartSeg/plugins/napari_widgets/measurement_widget.py#L24)
* [napari-stracking `STracksFeatures`](https://github.com/sylvainprigent/napari-stracking/blob/main/napari_stracking/_sfeatures_plugins.py#L6)
* [napari-features `features()`](https://github.com/0x00b1/napari-features/blob/master/napari_features/_features.py#L11)

Anyone with write-access, please feel free to edit this issue and add other items to this list.

Best,
Robert

**Comment by kevinyamauchi** [comment]
Hey @haesleinhuepf! Thanks for gathering so many related plugins. I have also been thinking about this and I am hesitant to add yet another features plugin. I made this repository mostly to get the thoughts down from our meeting and start this conversation. My vision for `morphometrics` is more about _using_ features measurements to explore tissue structure and segmentation quality than _making_ features measurements. In that lens, I don't have a strong preference for interop vs incorporation for `morphometrics` itself.

In my mind, the unifying features plugin would have a solid set of "core" measurements as well as the ability for other plugins to contribute measurement functions via something like the `register_measurement()` decorator in #1. The reason for the core measurements is that I think it is nice to solid implementations of the main measurements in one place (e.g., caching as suggested in #1). The ability to register measurements from other libraries is nice because it makes it easy to optionally include measurements from libraries with tricky dependencies (e.g., opencl, cuda).

In terms of the existing plugins, I like all of them. However, I am not sure that `napari-features` is being actively maintained and the partseg one is part of the larger PartSeg library (as I understand it).

I am wondering if `napari-clusters-plotter` can become the "main" features plugin that we all contribute to. It already has nice visualization for the measurements and I think it is nice to include a couple of built in dimensionality reduction techniques.

A few questions related questions:
- Do you think `napari-skimage-regionprops` can be incorporated into `napari-cluster-plotter` and "deprecated"? Or are there reasons both should continue to be maintained?
- Would you and @lazigu be open to a PR that adds the ability to register measurement functions/classes to `napari-cluster-plotter`?
- Can the cupy and pyclesperanto libraries register their measurement functions to the "main" features plugin?
- If yes to the above questions, do you think `napari-cluster-plotter` can become the  unifying features plugin?

What do you all think? Does this make any sense?

**Comment by kevinyamauchi** [comment]
As a side note, I think it generally makes sense for these "common" applications to have a community has a core piece that individual developers build upon. The core piece reduces the redundant work and helps with interoperability and each developer is able to move quickly on their own application. I think trying to build applications directly on the core piece feels slow and hard to experiment because each addition requires community buy in. This is my motivation for wanting to create `morphometrics` instead of purely trying to build the application by purely contribute directly to an existing plugin like `napari-clusters-plotter`.

**Comment by haesleinhuepf** [comment]
TL;DR: What do you think about sending the `register_measurement` to napari-skimage-regionprops at making it the "main" features plugin?

> My vision for morphometrics is more about using features measurements to explore tissue structure and segmentation quality than making features measurements.

Ah, I see. So I had clearly a misunderstanding here. Thanks for clarifying!

> * Do you think `napari-skimage-regionprops` can be incorporated into `napari-cluster-plotter` and "deprecated"? Or are there reasons both should continue to be maintained?
> * Would you and @lazigu be open to a PR that adds the ability to register measurement functions/classes to `napari-cluster-plotter`?

Very interesting questions! napari-skimage-regionprops is a lightweight-dependency. Hence, it installs on all computers easily. It is also used from [some other plugins](https://github.com/haesleinhuepf/napari-skimage-regionprops/network/dependents), and hence, it already _is_ a core piece from my perspective. It also comes with [basic infrastructure for handling tables](https://github.com/haesleinhuepf/napari-skimage-regionprops/blob/master/napari_skimage_regionprops/_table.py#L139) (GUI) as long as this is not available in napari-core. The napari-clusters-plotter on the other side is not so lightweight and is tricky to install. It needs hdbscan and pyopencl, which don't come without any trouble to any machine (via pip). 

> * Can the cupy and pyclesperanto libraries register their measurement functions to the "main" features plugin?

Sure! That makes perfect sense.

> If yes to the above questions, do you think napari-cluster-plotter can become the unifying features plugin?

We need to resolve the dependency issue first. Maybe that comes for free as soon as @goanpeca finishes his work on making all napari-plugins conda-installable? At the moment, I would prefer to collect things in napari-skimage-regionprops.

**Comment by kevinyamauchi** [comment]
Thanks for the quick response, @haesleinhuepf !

> TL;DR: What do you think about sending the register_measurement to napari-skimage-regionprops at making it the "main" features plugin?

Based on your explanation about the dependencies of `napari-skiamge-regionprops` and `napari-clusters-plotter`, this makes total sense. A couple of follow-up questions:

- Are you open to renaming `napari-skimage-regionprops` to `napari-regionprops` or some other more general name?
- Should we have discussions about function signatures and types for the measurements on the `napari-skimage-regionprops` repo?

**Comment by haesleinhuepf** [comment]
> Are you open to renaming napari-skimage-regionprops to napari-regionprops or some other more general name?

No strong opinion, but renaming a package essentially means creating a new one, right? We can totally do this, we just need to be a bit careful. napari-skimage-regionpros is downloaded a couple of hundred times a month. We shouldn't break those user's napari during this renaming.

> Should we have discussions about function signatures and types for the measurements on the napari-skimage-regionprops repo?

Yes, that's very welcome!
